src/contracts/core/AVSDirectory.sol

```diff
@@ -581,42 +586,40 @@ contract AVSDirectory is
             // 3. ensure only queued deallocation per operator, operatorSet, strategy
             _checkPendingDeallocations(operator, strategy, operatorSets[i]);
 
-            // 4. queue deallocation for (op, opSet, strategy) for magnitudeDiff amount
-            _queuedDeallocations[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId].push(
-                QueuedDeallocation({
+            // 4. push PendingFreeMagnitude and respective array index into (op,opSet,Strategy) queued deallocations
+            uint256 index = _pendingFreeMagnitude[operator][strategy].length;
+            _pendingFreeMagnitude[operator][strategy].push(
+                PendingFreeMagnitude({
                     magnitudeDiff: deallocation.magnitudeDiffs[i],
                     completableTimestamp: completableTimestamp
                 })
             );
+            _queuedDeallocationIndices[operator][strategy][operatorSets[i].avs][operatorSets[i].operatorSetId].push(index);
         }
     }
 
-    /**
-     * @notice Completes queued deallocations for a single strategy and multiple operator sets
-     * @param operator address to complete deallocations for
-     * @param strategy the strategy to complete deallocations for
-     * @param operatorSet the operator set to complete deallocations for the given strategy
-     */
-    function _completeDeallocation(
+    /// @dev read through pending free magnitudes and add to freeMagnitude if completableTimestamp is >= block timestamp
+    /// In additiona to updating freeMagnitude, updates next starting index to read from for pending free magnitudes after completing
+    function _updateFreeMagnitude(
         address operator,
```

Comment:
probably want to set an `endBeforeIndex` here that is ignored if set to `0`

Reply:
Think we are okay with just completing all. This will be called in both allocate/deallocate functions to bound the "numToComplete" from increasing past 1.

Reply:
Synced offline but there is a bound on (op, strategy, opSet) queuedDeallos but not on pendingFreeMagnitude(op,strategy) because multiple opSets can be queued and not completed

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/680/commits/c3d36fdef27069dc8b643ecba7b638963756a7a6